dali/kernels/slice/slice_flip_normalize_permute_pad_cpu.h

```diff
@@ -30,165 +31,291 @@ namespace kernels {
+    destination = ConvertSat<OutputType>(fpout);
+  } else {
     (void) mean;
     (void) inv_stddev;
-    if (std::is_integral<OutputType>::value && std::is_floating_point<InputType>::value) {
-      destination = clamp<OutputType>(std::roundf(element));
+    destination = ConvertSat<OutputType>(element);
+  }
+}
+
+template <bool NeedNormalize, bool HasChannels, bool OutOfBounds, typename OutputType,
+          typename InputType>
+void SliceFlipNormalizePermutePadKernelImpl(
+    OutputType *output, const InputType *input, const int64_t *in_strides,
+    const int64_t *out_strides, const int64_t *anchor, const int64_t *in_shape,
+    const int64_t *out_shape, const OutputType *fill_values, const float *mean,
+    const float *inv_stddev,
+    int channel_dim,  // negative if no channel dim or already processed
+    std::integral_constant<int, 1>) {
+  constexpr int d = 0;
+  if (OutOfBounds) {
+    if (HasChannels && d == channel_dim) {
+      for (int i = 0; i < out_shape[d]; i++) {
+        output[i] = *fill_values;
+        fill_values++;
+        mean++;
+        inv_stddev++;
+      }
     } else {
-      destination = clamp<OutputType>(element);
+      for (int i = 0; i < out_shape[d]; i++) {
+        output[i] = *fill_values;
+      }
     }
-  }
-};
+  } else  {
+    int64_t pad_before, slice, pad_after;
+    std::tie(pad_before, slice, pad_after) =
+        CalcPadCopyExtents(anchor[d], in_shape[d], out_shape[d]);
+    if (in_strides[d] < 0) std::swap(pad_before, pad_after);
 
-struct NormalizePolicy {
-  template <typename OutputType, typename InputType>
-  static inline void Fill(OutputType &destination, InputType element,
-                          const float *mean, const float *inv_stddev) {
-    float fpout = (static_cast<float>(element) - (*mean)) * (*inv_stddev);
-    if (std::is_integral<OutputType>::value) {
-      destination = clamp<OutputType>(std::roundf(fpout));
+    // out of bounds (left side)
+    if (HasChannels && d == channel_dim) {
+      for (int64_t i = 0; i < pad_before; i++) {
+        *output = *fill_values;
+        output += 1;  // out_strides[d] is 1;
+        input  += in_strides[d];
+        fill_values++;
+        mean++;
+        inv_stddev++;
+      }
     } else {
-      destination = clamp<OutputType>(fpout);
+      for (int64_t i = 0; i < pad_before; i++) {
+        *output = *fill_values;
+        output += 1;  // out_strides[d] is 1;
+        input  += in_strides[d];
+      }
     }
-  }
-};
 
-template <typename OutputType, int Dims>
-inline void Pad(OutputType *output,
-                TensorShape<Dims> out_strides,
-                TensorShape<Dims> padded_out_shape,
-                float fill_value,
-                std::integral_constant<int, 1>) {
-  constexpr auto d = Dims - 1;
-  for (int64_t i = 0; i < padded_out_shape[d]; i++) {
-    *output = Convert<OutputType>(fill_value);
-    output += out_strides[d];
+    // within input bounds
+    if (HasChannels && d == channel_dim) {
```

Comment:
Maybe `SliceFlipNormalizePermuteKernelImpl` can be used to fill `//within input bounds`?

Reply:
Not really, the next dimension might need padding